File: src/cpp_pubsub_srv/src/listener_adv.cpp
#include <memory>
#include <chrono>
#include <functional>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "custom_interfaces/msg/location.hpp"
#include "custom_interfaces/srv/authenticate_listener.hpp"
using std::placeholders::_1;

class AdvancedListener : public rclcpp::Node
{
  public:
    AdvancedListener()
    : Node("adv_listener")
    {
      subscription_ = this->create_subscription<custom_interfaces::msg::Location>(
      "custom_topic_adv", 10, std::bind(&AdvancedListener::topic_callback, this, _1));
      publisher_ = this->create_publisher<custom_interfaces::msg::Location>("share_secret", 10);

      client_ = this->create_client<custom_interfaces::srv::AuthenticateListener>("/adv_talker/share_secret_location");

      this->declare_parameter("share_parameter", true);

      timer_ = this->create_wall_timer(std::chrono::seconds(2), std::bind(&AdvancedListener::call_location_service, this));
    }

  private:
    void topic_callback(const custom_interfaces::msg::Location & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "I heard: (%f, %f, %f)/%s", 
                      msg.location.x, msg.location.y,
                      msg.location.z, msg.name.c_str());
    }

    void call_location_service()
    {
      if (!client_->wait_for_service(std::chrono::seconds(2))) {
        RCLCPP_WARN(this->get_logger(), "Service not available...");
        return;
    }

      auto request = std::make_shared<custom_interfaces::srv::AuthenticateListener::Request>();
      request->password = 12345;

      auto future = client_->async_send_request(request,
        std::bind(&AdvancedListener::handle_location_response, this, std::placeholders::_1));

    }

    void handle_location_response(rclcpp::Client<custom_interfaces::srv::AuthenticateListener>::SharedFuture future)
    {
      auto response = future.get();
      
      RCLCPP_INFO(this->get_logger(),
                  "Authenticated location: (%f, %f, %f)/%s",
                  response->secret_location.location.x,
                  response->secret_location.location.y,
                  response->secret_location.location.z,
                  response->secret_location.name.c_str());
      
      const auto & loc = response->secret_location;

      if(this->get_parameter("share_parameter").as_bool())
        publisher_->publish(loc);
    }


  rclcpp::Subscription<custom_interfaces::msg::Location>::SharedPtr subscription_;
  rclcpp::Publisher<custom_interfaces::msg::Location>::SharedPtr publisher_;
  rclcpp::Client<custom_interfaces::srv::AuthenticateListener>::SharedPtr client_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AdvancedListener>());
  rclcpp::shutdown();
  return 0;
}

File: src/py_pubsub_srv/py_pubsub_srv/listener_adv.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from custom_interfaces.msg import Location
from custom_interfaces.srv import AuthenticateListener

class AdvancedListener(Node):
    def __init__(self):
        super().__init__('adv_listener')

        # Subscriber
        self.subscription_ = self.create_subscription(
            Location,
            'custom_topic_adv',
            self.topic_callback,
            10
        )
        self.subscription_  # prevent unused variable warning

        # Publisher
        self.publisher_ = self.create_publisher(Location, 'share_secret', 10)

        # Client
        self.client_ = self.create_client(AuthenticateListener, '/adv_talker/share_secret_location')

        # Parameter
        self.declare_parameter('share_parameter', True)

        # Timer (2 seconds)
        self.timer_ = self.create_timer(2.0, self.call_location_service)

        self.get_logger().info("AdvancedListener node started.")

    def topic_callback(self, msg):
        self.get_logger().info(
            f"I heard: ({msg.location.x:.2f}, {msg.location.y:.2f}, {msg.location.z:.2f})/{msg.name}"
        )

    def call_location_service(self):
        if not self.client_.wait_for_service(timeout_sec=2.0):
            self.get_logger().warn("Service not available...")
            return

        request = AuthenticateListener.Request()
        request.password = 12345

        future = self.client_.call_async(request)
        future.add_done_callback(self.handle_location_response)

    def handle_location_response(self, future):
        try:
            response = future.result()
        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")
            return

        loc = response.secret_location
        self.get_logger().info(
            f"Authenticated location: ({loc.location.x:.2f}, {loc.location.y:.2f}, {loc.location.z:.2f})/{loc.name}"
        )

        # Publish secret if parameter is True
        if self.get_parameter('share_parameter').get_parameter_value().bool_value:
            self.publisher_.publish(loc)
            self.get_logger().info("Published secret location to /share_secret")

def main(args=None):
    rclpy.init(args=args)
    node = AdvancedListener()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Shutting down AdvancedListener node.")
    finally:
        node.destroy_node()
        rclpy.shutdown()